### DISCUSSION TITLE
How to modify response from workflow instances journal activity data??

### DISCUSSION THREAD

**Discussion by tannontawat** [elsa-workflows/elsa-core#1577] [Q&A] [answered]
When I want to see the workflow instances activity data example inbound request in HTTP endpoint it have some sensitive data. So I want to sensor it before response (not modify it on WorkflowExecutionLogRecords table). What the best way to do it??

Best regards.

**Comment by sfmskywalker** [comment]
You can't easily do it right now, but it makes sense to enable this.
We need a way to allow your application code to mask PII away. Some users will want to mask this data before it gets persisted while others might want to persist the original data, but only mask it to dashboard users.

To mask data before it gets persisted, we could introduce a filter that you implement.
For example, you would implement something like this:

```csharp
public class PIIFilter : IWorkflowJournalFilter
{
   public void Apply(WorkflowJournalFilterContext context)
   {
      var activityTypeName = context.ActivityBlueprint.Type;
      var journalData = context.Data; // This is of type JObject.

      // Ignore everything that does not concern HTTP Endpoints.
      if(activityTypeName != nameof(HttpEndpoint))
         return data;
     
      var httpRequestModel = journalData.GetValue<HttpRequestModel>("Inbound Request"); // Get the stored HTTP request model.

      // Mask aspects of the HTTP request.
      httpRequestModel.RawBody = Mask(httpRequestModel.RawBody);
      httpRequestModel.Body = Mask(httpRequestModel.Body);

      journalData["Inbound Request"] = JObject.FromObject(httpRequestModel);
      return;

     // Alternatively, you could concert the HTTP request model to a JObject and remove certain parts entirely:
     var httpRequestModelJObject = JObject.FromObject(httpRequestModel);

     // Remove the Body field.
     httpRequestModelJObject.Remove("Body");

     // Mask the RawBody field (perhaps easier to process because it's a plain string).
     httpRequestModelJObject["RawBody"] = Mask(httpRequestModelJObject["RawBody"]);
   }
}
```

Would something like that work for your use case?

**Reply by tannontawat** [reply]
- What namespace should I import for IWorkflowJournalFilter and WorkflowJournalFilterContext?
- After I write this class, I must to register this class on Startup.cs or not ?

**Reply by sfmskywalker** [reply]
It doesn't exist today, I was proposing the above as a design for a solution. I'll create an issue for this as a new feature.

**Comment by tannontawat** [comment, marked as answer]
Thank you for your suggestion.,

Now, I use `IActionFilter` to process the executed event (`OnActionExecuted`) and register it in global config. I modify the data from 2 requests
- `Elsa.Server.Api.Endpoints.WorkflowInstances.Get` -> to get workflow instance
- `Elsa.Server.Api.Endpoints.WorkflowExecutionLog.Get` -> to get workflow execution log

because the workflow instance dashboard will call data from above Api (management Api).
Now, I usually modify data from `HttpRequestModel` and raw content (json string).
